src/javascript/url-shortener.js

```diff
@@ -12,71 +12,327 @@
+      // If it wasn't a forbidden error, return it to the caller. Otherwise,
+      // it's possible we have a bad token and should retry with a fresh one.
+      if (e !== forbiddenError) {
+        throw e;
+      }
+
+      // Hmm. I guess the token was bad. Clear the saved stuff before
+      // proceeding.
+      removeToken();
+      localStorage.removeItem(BITLY_GUID_STORAGE_KEY);
+    }
+  }
+
+  // Okay, so we have no token. Send the user to the bitly oauth url.
+
+  // This client ID should be attached as a global by the server template
+  // renderer. This way we don't have to make a whole round trip to an API to
+  // get it. This currently happens inside of lib/template.py:render, so it's
+  // always available.
+  const bitlyClientId = window.BITLY_CLIENT_ID;
+  if (!bitlyClientId) {
+    throw new Error(
+      'No OAuth client ID available. Make sure to attach it as a ' +
+      'global in the server!'
+    );
+  }
+
+  // Initiate OAuth flow. Spawn a new window, which sends the user to bitly's
+  // auth page. That will redirect us to /url-shorten/auth-callback, which
+  // will get an authentication token, then send us a message via postMessage
+  // with the token.
+  //
+  // We can't use try-finally here because the various APIs we're using (
+  // addEventListener, timers, etc) aren't promisified or awaitable, so we have
+  // to wrap them up in a promise abstraction that cleans up after itself.
+  //
+  // TODO(nathanwest): RxJS was a relatively late addition to the URL shortener
+  // design, but now that we have it, see if any of this cleanup stuff can be
+  // more easily or cleanly expressed using that library.
+  const {token} = await cleanupingPromise((resolve, reject, cleanup) => {
+    // Attach a listener that will await a message via postMessage
+    const messageListener = event => {
+      if (
+        event.origin === window.location.origin &&
+        event.source === childWindow
+      ) {
+        const data = event.data;
+        if (data.error) {
+          reject(new Error(data.error));
+        } else {
+          resolve(data);
+        }
+      }
+    };
+    window.addEventListener('message', messageListener, {once: true});
+    cleanup(() => window.removeEventListener('message', messageListener));
+
+    const authUrl = 'https://bitly.com/oauth/authorize' + encodeQuery({
+      client_id: bitlyClientId,
+      // The auth callback does about half the work here: it receieves an
+      // authorization code, converts it into a token, then sends the token
+      // back to us. This happens server side because it requires the
+      // client_secret, which we don't want to share with our users.
+      redirect_uri: window.location.origin + '/bitly-api-token-handler/',
+    });
+    // Open a new window to do authentication.
+    const childWindow = window.open(authUrl, '_blank');
+    if (childWindow == null) {
+      throw new Error(
+        'Failed open new window for authorizing bitly. ' +
```

Comment:
Phrasing here is a bit odd. Could you reword?

Reply:
Done